network/src/connectivity_manager/mod.rs

```diff
@@ -438,6 +464,42 @@ fn log_dial_result(peer_id: PeerId, addr: Multiaddr, dial_result: DialResult) {
     }
 }
 
+impl Addresses {
+    fn new() -> Self {
+        Default::default()
+    }
+
+    fn from_addrs(src: DiscoverySource, src_addrs: Vec<Multiaddr>) -> Self {
+        let mut addrs = Self::new();
+        addrs.update(src, src_addrs);
+        addrs
+    }
+
+    fn len(&self) -> usize {
+        self.0.iter().map(Vec::len).sum()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn update(&mut self, src: DiscoverySource, addrs: Vec<Multiaddr>) {
+        let idx = src as u8 as usize;
+        self.0[idx] = addrs;
+    }
+
+    fn get(&self, mut idx: usize) -> Option<&Multiaddr> {
+        for addrs in &self.0 {
```

Comment:
might be able to simplify using: `self.0.iter().as_ref().flatten().nth(idx)`

Reply:
might even be able to do it without `.by_ref()`

Reply:
damn, that's pretty big brain. looks like it works too